### PR TITLE
os/bluestore: separate kv_sync_thread into two parts

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

**Comment by markhpc** [comment]
This is failing to compile both in jenkins and on incerta in the same way (see the jenkins log).  After some digging I suspect this might be due to gcc 4.8 not liking the c++11x syntax used here.  If it compiles with gcc >= 5.x we may need to rewrite that part of the code if we care about 4.8 compatability (ie centos7).

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
sage-2017-03-25_14:58:48-rados-wip-bluestore-kv-finisher---basic-smithi

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -7460,20 +7460,20 @@ void BlueStore::_txc_committed_kv(TransContext *txc)
     txc->onreadable_sync->complete(0);
     txc->onreadable_sync = NULL;
   }
-  unsigned n = txc->osr->parent->shard_hint.hash_to_shard(m_finisher_num);
   if (txc->oncommit) {
     logger->tinc(l_bluestore_commit_lat, ceph_clock_now() - txc->start);
-    finishers[n]->queue(txc->oncommit);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Shouldn't we remove 'finishers' stuff from BLueStore at all?

**Comment by liewegas** [comment]
Yeah, I just started to doubt whether we want to go all the way down this 
path since we're basically replacing code that pushes work to finisher(s) 
with code that pushes work to one finalizer.  Maybe we should see what the 
qd=1 and/or single osd tests from @markhpc show next?

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
This deadlocks when run under a loaded osd.  the core files are jacked up, but the last line i see for the finalizer thread is 
<pre>
2017-04-13 19:36:55.886926 7f33cd9d5700 20 bluestore(/var/lib/ceph/osd/ceph-4) _txc_committed_kv txc 0x562da76ffb00
</pre>
which suggests it's deadlocking while doing the completions.

**Comment by liewegas** [comment]
That was from this job: http://pulpito.ceph.com/sage-2017-04-13_19:24:30-rados-wip-sage-testing---basic-smithi/1022074/

**Comment by markhpc** [comment]
See #15163

**Comment by liewegas** [comment]
tabling this approach until we resolve the deadlock between split's use of osr_drain (while holding pg lock) vs the completions that may also block on pg lock.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -8128,11 +8135,83 @@ void BlueStore::_kv_sync_thread()
...
+      if (kv_stop)
+	break;
+      dout(20) << __func__ << " sleep" << dendl;
+      kv_sync_cond.notify_all();
+      kv_finalize_cond.wait(l);
+      dout(20) << __func__ << " wake" << dendl;
+    } else {
+      kv_committed.swap(kv_committing_to_finalize);
+      deferred_stable.swap(deferred_stable_to_finalize);
+      kv_finalize_cond.notify_one();
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: we don't need that IMO

File: src/os/bluestore/BlueStore.h [hunk 3]
```diff
@@ -1809,6 +1817,12 @@ class BlueStore : public ObjectStore,
   deque<DeferredBatch*> deferred_done_queue;   ///< deferred ios done
   deque<DeferredBatch*> deferred_stable_queue; ///< deferred ios done + stable
 
+  KVFinalizeThread kv_finalize_thread;
+  std::mutex kv_finalize_lock;
+  std::condition_variable kv_finalize_cond;
+  deque<TransContext*> kv_committing_to_finalize;   ///< pending finalization
+  deque<TransContext*> deferred_stable_to_finalize; ///< pending finalization
```
Comment (markhpc) [inline src/os/bluestore/BlueStore.h]: This needs to be:

deque<DeferredBatch*> deferred_stable_to_finalize;

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -8128,11 +8135,82 @@ void BlueStore::_kv_sync_thread()
...
+      // commit cycle.
+      deferred_stable_queue.swap(deferred_done);
+    }
+  }
+  dout(10) << __func__ << " finish" << dendl;
+}
+
+void BlueStore::_kv_finalize_thread()
+{
+  deque<TransContext*> deferred_stable, kv_committed;
```
Comment (markhpc) [inline src/os/bluestore/BlueStore.cc]: deque<DeferredBatch*> deferred_stable;
deque<TransContext*> kv_committed;

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -8128,11 +8135,82 @@ void BlueStore::_kv_sync_thread()
...
+  std::unique_lock<std::mutex> l(kv_finalize_lock);
+  while (true) {
+    assert(kv_committed.empty());
+    assert(deferred_stable.empty());
+    if (kv_committing_to_finalize.empty() &&
+	deferred_stable_to_finalize.empty()) {
+      if (kv_stop)
+	break;
+      dout(20) << __func__ << " sleep" << dendl;
+      kv_sync_cond.notify_all();
```
Comment (markhpc) [inline src/os/bluestore/BlueStore.cc]: This doesn't exist anymore, Igor got rid of it in https://github.com/ceph/ceph/commit/71d60ba8c4397cda584ae801a7542ea9ebde5069

**Comment by markhpc** [comment]
A quick performance test shows this improving on master in the expected range.  I think this is ready for qa-testing!

**Pull request merged**